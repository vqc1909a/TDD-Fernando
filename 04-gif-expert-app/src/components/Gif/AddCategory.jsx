import { useState } from "react"
import PropTypes from 'prop-types'


const DEFAULT_VALUE = "";
 
export const AddCategory = ({handleAddCategory}) => {
  
  const [inputValue, setInputValue] = useState(DEFAULT_VALUE);

  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  }

  const handleSubmit = (e) => { 
    e.preventDefault(); 
    if(inputValue.trim()){
      handleAddCategory(inputValue);
      setInputValue(DEFAULT_VALUE);
    }
  } 
  return (
    <form onSubmit={handleSubmit}>
        <input type="text" placeholder="Buscar gifts" value={inputValue} onChange={handleInputChange}/>
    </form>
  )
}

AddCategory.propTypes = {
  handleAddCategory: PropTypes.func.isRequired
}