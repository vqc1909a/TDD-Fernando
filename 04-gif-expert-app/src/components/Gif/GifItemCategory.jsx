import { PropTypes } from "prop-types"

export const GifItemCategory = ({title, url}) => {
  return (
    <div className="card" aria-label="card">
      <img src={url} alt={title}></img>
      <p>{title}</p>
    </div>
  )
}
GifItemCategory.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}

/* Tarea

1. AÃ±adir PropTypes a GifItemCategory
  a. title obligatorio
  b. url obligatorio

2. Evaluar el snapshot

*/
