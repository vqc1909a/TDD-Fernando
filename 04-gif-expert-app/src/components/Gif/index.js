//This is a file of exports. It is used to export all the components from the folder Gif, This is also called "barrel" file
export * from "./AddCategory";
export * from "./GifGridCategory";
export * from "./GifItemCategory";

