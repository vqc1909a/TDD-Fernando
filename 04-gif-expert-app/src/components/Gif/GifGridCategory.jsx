import { PropTypes } from "prop-types"
import { GifItemCategory } from "../../components/Gif";
import { useFetchGifs } from "../../hooks/useFetchGifs";

export const GifGridCategory = ({category}) => {  
  const { gifs, isLoading }  = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>  
      {
        isLoading
        ?
        <p>Loading...</p>
        :
        <div className="card-grid" aria-label="card-grif">
          {gifs.map(image => <GifItemCategory key={image.id} {...image}></GifItemCategory>)}
        </div>
      }
    </>
  )
}

GifGridCategory.propTypes = {
  category: PropTypes.string.isRequired
}

