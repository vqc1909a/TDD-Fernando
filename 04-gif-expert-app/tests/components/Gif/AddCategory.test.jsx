import {fireEvent, render, screen} from "@testing-library/react";
import { AddCategory } from "../../../src/components/Gif/AddCategory";

describe('tests on component AddCategory', () => { 
  test('component should make to match to the snapshot', () => { 
    const {container} = render(<AddCategory handleAddCategory={() => {}}></AddCategory>);
    expect(container).toMatchSnapshot();
  })

  test('should change the value input', () => { 
    let valueToChange = "One Punch";

    render(<AddCategory handleAddCategory={() => {}}></AddCategory>);
    let inputToChangeValue = screen.getByRole('textbox');
    expect(inputToChangeValue).toBeTruthy();

    //The difference between input and change is that the input event is fired when the value of an input changes, and the change event is fired when the value of an input changes and the input loses focus.

    //The second parameter is the value that we want to change
    fireEvent.change(inputToChangeValue, {target: {value: valueToChange}})
    expect(inputToChangeValue.value).toBe(valueToChange);
    // screen.debug();
  })

  test("should call the function handleAddCategory when let's trigger submit formulary with some value", () => { 
    let valueToChange = "One Punch";
    let handleAddCategory = jest.fn();

    render(<AddCategory handleAddCategory={handleAddCategory}></AddCategory>);
    let inputToChangeValue = screen.getByRole('textbox');
    let formToTriggerSubmit = screen.getByRole('form');
    fireEvent.change(inputToChangeValue, {target: {value: valueToChange}})
    
    //In the event submit, the second parameter is the event object, but we don't need it, so we pass an object with a method preventDefault that does nothing   
    fireEvent.submit(formToTriggerSubmit, {preventDefault(){}});
    expect(inputToChangeValue.value).toBe("");
    
    expect(handleAddCategory).toHaveBeenCalled();
    expect(handleAddCategory).toHaveBeenCalledTimes(1);
    expect(handleAddCategory).toHaveBeenCalledWith(valueToChange);
  })

  test("shouldn't call the function handleAddCategory when let's trigger submit formulary with empty value", () => { 
    let handleAddCategory = jest.fn();

    render(<AddCategory handleAddCategory={handleAddCategory}></AddCategory>);
    let formToTriggerSubmit = screen.getByRole('form');
    // screen.debug();

    //In the event submit, the second parameter is the event object, but we don't need it, so we pass an object with a method preventDefault that does nothing   
    fireEvent.submit(formToTriggerSubmit, {preventDefault(){}});
    expect(handleAddCategory).not.toHaveBeenCalled();
    expect(handleAddCategory).toHaveBeenCalledTimes(0);
  })
})

