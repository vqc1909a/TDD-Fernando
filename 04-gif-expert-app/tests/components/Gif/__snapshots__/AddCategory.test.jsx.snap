// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on component AddCategory component should make to match to the snapshot 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar gifts"
      type="text"
      value=""
    />
  </form>
</div>
`;
